json/src/value.rs

```diff
@@ -161,26 +168,97 @@ impl Value {
         }
+        Some(target)
+    }
+
+    /// Looks up a value by a JSON Pointer and returns a mutable reference to
+    /// that value.
+    ///
+    /// JSON Pointer defines a string syntax for identifying a specific value
+    /// within a JavaScript Object Notation (JSON) document.
+    ///
+    /// A Pointer is a Unicode string with the reference tokens separated by `/`.
+    /// Inside tokens `/` is replaced by `~1` and `~` is replaced by `~0`. The
+    /// addressed value is returned and if there is no such value `None` is
+    /// returned.
+    ///
+    /// For more information read [RFC6901](https://tools.ietf.org/html/rfc6901).
+    pub fn pointer_mut<'a>(&'a mut self, pointer: &str) -> Option<&'a mut Value> {
         if pointer == "" {
             return Some(self);
         }
         if !pointer.starts_with('/') {
             return None;
         }
+        let tokens = pointer.split('/').skip(1).map(|x| x.replace("~1", "/").replace("~0", "~"));
         let mut target = self;
-        for escaped_token in pointer.split('/').skip(1) {
-            let token = escaped_token.replace("~1", "/").replace("~0", "~");
-            let target_opt = match *target {
-                Value::Object(ref map) => map.get(&token[..]),
-                Value::Array(ref list) => {
-                    parse_index(&token[..]).and_then(|x| list.get(x))
-                }
+
+        for token in tokens {
+            // borrow checker gets confused about `target` being mutably borrowed too many times because of the loop
+            // this once-per-loop binding makes the scope clearer and circumvents the error
+            let target_once = target;
+            let target_opt = match *target_once {
+                Value::Object(ref mut map) => map.get_mut(&token),
+                Value::Array(ref mut list) => parse_index(&token).and_then(move |x| list.get_mut(x)),
+                _ => return None,
+            };
+            if let Some(t) = target_opt {
+                target = t;
+            } else {
+                return None;
+            }
+        }
+        Some(target)
+    }
+
+    /// Looks up a value by a JSON Pointer while consuming the object to return
+    /// the value as owned, immutable data.
+    ///
+    /// JSON Pointer defines a string syntax for identifying a specific value
+    /// within a JavaScript Object Notation (JSON) document.
+    ///
+    /// A Pointer is a Unicode string with the reference tokens separated by `/`.
+    /// Inside tokens `/` is replaced by `~1` and `~` is replaced by `~0`. The
+    /// addressed value is returned and if there is no such value `None` is
+    /// returned.
+    ///
+    /// For more information read [RFC6901](https://tools.ietf.org/html/rfc6901).
+    pub fn pointer_owned(self, pointer: &str) -> Option<Value> {
+        if pointer == "" {
+            // except for this, the function could probably be `&mut self` and only consume the child element
```

Comment:
I would prefer to keep it the way @CryptArchy has written it. The behavior you suggested is `v.pointer_mut(p).map(|m| mem::replace(m, Value::Null))` which makes it clear in the caller's code exactly how the value is being mutated.